Blocks: Add Facebook and Instagram Embed block variations

ockham opened the pull request:
"Successor" to D49605-code and 522-gh-wpcomsh. Once this is merged to JP, and the corresponding WP.com diff (D49756-code) is committed to WP.com, we can retire those changes.

#### Changes proposed in this Pull Request:

Quoting 522-gh-wpcomsh:

> Register a callback for the `block.registerBlockType` frontend filter that patches the `Facebook` and `Instagram` `core/embed` block variations so they appear in the inserter and can have their relevant URLs parsed again.
>
> This is relevant for **Gutenberg >= 9.0** (see [CfT](p7DVsv-9jU-p2#comment-31039)), where the aforementioned blocks [were deprecated](https://github.com/WordPress/gutenberg/pull/24472) because they do not support the oEmbed changes that Facebook is going to implement on Oct. 24th, 2020, and core has currently no plans to support it.
>
>  However, we do plan on keeping support for them on WPCOM/AT. Facebook is adding a token authentication mechanism to its new oEmbed implementation, and we are following suit.
>
>Our goal is for this go unnoticed by our end-users, as this is only an implementation detail.

#### Jetpack product discussion
p7DVsv-9jU-p2#comment-31116

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
Make sure you're testing on a WP install with Gutenberg 9.0 installed.

This diff is rather simple -- the crucial part is to make sure that we're actually not making any more requests to the legacy API endpoints for Instagram oEmbeds (https://api.instagram.com/oembed). We'll be testing that quite thoroughly in the following.

The Facebook embeds OTOH are largely unaffected, since they haven't been previewed in the editor before (instead, we've been displaying an `Embedded content from facebook.com can't be previewed in the editor.` message in the block). (Some discussion on whether that's the way it _should_ be here: https://github.com/Automattic/wp-calypso/issues/27701#issuecomment-693351019, but that's not really relevant to this PR.)

- We need to direct the REST API requests your local Jetpack dev install makes to your WP.com sandbox: In your local Jetpack dev env, sandbox the REST API, e.g. by adding a `0-sandbox.php` file to `docker/mu-plugins/`, with the following contents:

```php
<?php

define( 'JETPACK__SANDBOX_DOMAIN', 'YOURSANDBOX' );
```

(change `YOURSANDBOX` to your sandbox' URL). Make sure to restart your JP install afterwards. For more information on this, see PCYsg-efV-p2 (Section "The `JETPACK__SANDBOX_DOMAIN` PHP Constant").

- Make sure that your local JP install is really using your sandbox for these requests: Log out of your sandbox, and paste the link of an Instagram post into your JP install's Gutenberg editor. You should get an empty embed block showing a notice saying something like "Sorry, this content could not be embedded".
- Log into your sandbox, and verify that pasting an IG post URL is now successfully transformed into an embed. _Use a different IG post to avoid getting a cached result._ 
- Apply the following patch to your sandbox to [simulate](https://developers.facebook.com/docs/instagram/oembed-legacy) the retiring of Instagram's legacy API endpoint:

```diff
Index: wp-content/rest-api-plugins/endpoints/oembed-proxy.php
===================================================================
--- wp-content/rest-api-plugins/endpoints/oembed-proxy.php	(revision 213803)
+++ wp-content/rest-api-plugins/endpoints/oembed-proxy.php	(working copy)
@@ -52,8 +52,10 @@
 				'https://graph.facebook.com/v5.0/instagram_oembed'
 			);
 		} else {
+			$query_args = $request->get_query_params();
+			$query_args['breaking_change'] = 'oembed';
 			$request_url = add_query_arg(
-				$request->get_query_params(),
+				$query_args,
 				'https://api.instagram.com/oembed/'
 			);
 		}
```
- Try inserting another IG post URL (again a different one, because caching). It should fail.
- Fortunately, we already have code in place to use the new oEmbed API instead. To enable it, set

```
$auth_to_instagram = true;
```

on L40 of `wp-content/rest-api-plugins/endpoints/oembed-proxy.php` in your sandbox.
(I can't post a diff here, since it would include an internal link.)

- Verify that the block embed works now.
- Finally, we need to simulate WP.com behavior after the legacy oEmbed API has been retired. To that end, apply the following patch _to your local JP install_ (and reload the block editor):

```diff
diff --git a/modules/shortcodes/instagram.php b/modules/shortcodes/instagram.php
index 8a24c479b..bb80ed709 100644
--- a/modules/shortcodes/instagram.php
+++ b/modules/shortcodes/instagram.php
@@ -248,7 +248,8 @@ function jetpack_instagram_use_cache( $matches, $atts, $passed_url ) {
  * @return mixed An object if successful or a WP_Error object
  */
 function jetpack_instagram_fetch_embed( $args ) {
-       if ( Constants::is_defined( 'IS_WPCOM' ) && Constants::get_constant( 'IS_WPCOM' ) ) {
+       if ( true ) {
+               $args['breaking_change'] = 'oembed';
                $url      = esc_url_raw(
                        add_query_arg(
                                $args,

```
- Verify that pasting an IG URL doesn't work now. This means that we'll have to modify the `if` clause to make sure we're also using the new API for WP.com after Oct 24.
- Remove all patches that you've applied on top of this branch, or on your sandbox. Disconnect your JP from WP.com (at `/wp-admin/admin.php?page=jetpack#/dashboard`, "Manage website connection"). Turn on [Offline Mode](https://jetpack.com/support/development-mode/). Reload the editor, and verify that the IG embed is gone from the inserter, while the FB one is still there.

#### Conclusion from Testing Instructions

- The blocks added by this PR currently work -- the PR can be merged as-is :heavy_check_mark: 
- To prepare for Facebook's retiring of the old oEmbed API, we need to set `$auth_to_instagram = true` in `wp-content/rest-api-plugins/endpoints/oembed-proxy.php` before Oct 24. (We know this much is `true`.)
- To also make sure that the block variations work on WP.com after Oct 24, we need to file a PR that modifies the conditional in `jetpack_instagram_fetch_embed` in [`modules/shortcodes/instagram.php`](https://github.com/Automattic/jetpack/blob/8f8104179df236101209af8f71aef3e70d06d340/modules/shortcodes/instagram.php) accordingly.

#### Proposed changelog entry for your changes:
Add Facebook and Instagram Embed block variations.

cc/ @Automattic/good-mountain @ebinnion @jeherve

Fixes https://github.com/Automattic/wp-calypso/issues/45832

ebinnion commented on extensions/blocks/embed/editor.js:
```diff
@@ -0,0 +1,55 @@
+	 * made in https://github.com/WordPress/gutenberg/pull/24472.
+	 */
+	settings.variations.forEach( function ( variation ) {
+		switch ( variation.name ) {
```
I know this is a draft PR, so feel free to ignore this if you're already considering it. When I look at core, I see that there is a bit more information with each embed block.

<img width="962" alt="Screen Shot 2020-09-17 at 12 23 44 PM" src="https://user-images.githubusercontent.com/1126811/93505701-d057d000-f8e0-11ea-9ebf-e2b228d750b8.png">

Will we also need to add this information to this PR?

ockham replied:
That information should still be present. We're only modifying the FB and IG block variations (that [still exist](https://github.com/WordPress/gutenberg/blob/b34113c6d5eda5c459070c68179b2ed458567e60/packages/block-library/src/embed/variations.js#L58-L85) in GB) to change back what was modifed by https://github.com/WordPress/gutenberg/pull/24472 (which is what removed them from the block inserter, and made them not appear when pasting a matching FB or IG URL) :slightly_smiling_face:

ebinnion replied:
Got ya. Thanks for the clarification!

jeherve reviewed (commented):
It may be worth moving this to `extensions/extended-blocks`, since we're extending a core block (vs. creating our own block).

See #17137 for an example.

ockham commented:
> It may be worth moving this to `extensions/extended-blocks`, since we're extending a core block (vs. creating our own block).
> 
> See #17137 for an example.

Ah thanks, good point!

ockham commented:
> It may be worth moving this to `extensions/extended-blocks`, since we're extending a core block (vs. creating our own block).

f6216e60d

jeherve commented on extensions/extended-blocks/core-embed/index.js:
```diff
@@ -0,0 +1,55 @@
+ * @returns {object} The settings for the given block with the patched variations.
+ */
+function reactivateFacebookAndInstagramEmbedBlockVariations( settings, name ) {
+	if ( name !== 'core/embed' || ! settings.variations ) {
```
Do we need to account for regular WordPress 5.4 and 5.5 installations, where [this change](https://github.com/WordPress/gutenberg/pull/24090) hasn't happened yet and embed blocks are a bit different (`core-embed/facebook` I think?)?

ockham replied:
I tried to think this through, and I _think_ we don't have to do anything about them: Basically, they'll continue to exist in those versions, so unlike GB >= 9.0, nothing is being removed that we have to add back.

Does that make sense? :thinking:

jeherve replied:
It does, you're right! 👍

sirreal commented on extensions/extended-blocks/core-embed/index.js:
```diff
@@ -0,0 +1,55 @@
+	 * made in https://github.com/WordPress/gutenberg/pull/24472.
+	 */
+	settings.variations.forEach( function ( variation ) {
+		switch ( variation.name ) {
+			case 'facebook':
+				variation.patterns = [ /^https?:\/\/www\.facebook.com\/.+/i ];
+				delete variation.scope;
+				break;
+			case 'instagram':
+				variation.patterns = [ /^https?:\/\/(www\.)?instagr(\.am|am\.com)\/.+/i ];
+				delete variation.scope;
+				break;
+		}
```
Let's make sure we add these embeds in case they are ever _removed_ from core. See https://github.com/Automattic/wp-calypso/issues/45832

ockham replied:
2dda069e90316699c30d4281aa0381dae7264a2d

jeherve commented:
I may be missing something here, but since the embed registration will also be removed from Core in 50861-core, won't we be creating issues on sites where Jetpack's Shortcodes module (where both Facebook and Instagram embeds are registered) is deactivated? On those sites, we'll show the 2 embed variations, but won't have the embed registration happening behind the scenes.

This may happen on a lot of sites since the module is not activated by default.

retrofox commented on extensions/extended-blocks/core-embed/index.js:
```diff
@@ -0,0 +1,55 @@
+}
+
+addFilter(
+	'blocks.registerBlockType',
```
could we use `registerBlockVariation()` as default and use the filter method as a fallback?

ockham commented:
> I may be missing something here, but since the embed registration will also be removed from Core in [50861-core](https://href.li/?https://core.trac.wordpress.org/ticket/50861), won't we be creating issues on sites where Jetpack's Shortcodes module (where both Facebook and Instagram embeds are registered) is deactivated? On those sites, we'll show the 2 embed variations, but won't have the embed registration happening behind the scenes.
> 
> This may happen on a lot of sites since the module is not activated by default.

That's a great point, thanks!

Are we exposing the activation status of JP modules through a REST API endpoint already? (Can't remember off the top of my head :thinking: ) Otherwise I'll need to expose this some other way...

jeherve commented:
> Are we exposing the activation status of JP modules through a REST API endpoint already? (Can't remember off the top of my head 🤔 ) Otherwise I'll need to expose this some other way...

There are a few endpoints you could look at, and it may be worth creating a new shared utility in `extensions/shared` to fetch a module and its status:
https://github.com/Automattic/jetpack/blob/d8fe42cde874e5ef2019ca4ad33491b9a3045e92/_inc/lib/class.core-rest-api-endpoints.php#L274-L310

Or maybe we could pass the list of modules and their status here:
https://github.com/Automattic/jetpack/blob/96065a591242048cfa3b32697eae651b5ae63fbf/class.jetpack-gutenberg.php#L705

ockham commented on extensions/extended-blocks/core-embed/facebook.js:
```diff
@@ -0,0 +1,83 @@
+		// settings.variations = [ ...variations, facebookVariation ];
+
+		// Alas, it doesn't (yet), so we have to do it like this:
+		unregisterBlockVariation( 'core/embed', facebookVariation.name );
+		registerBlockVariation( 'core/embed', facebookVariation );
```
These happen async, whenever `apiFetch` is done. Fortunately, this seems to work pretty reliably, since `[un]registerBlockVariation` dispatches actions to [change state](https://github.com/WordPress/gutenberg/blob/ed848a0ac49257c1c8162fef09440a36c48a8375/packages/blocks/src/store/reducer.js#L146-L165) in the `blocks` store (basically using the variation name as a key). It appears that that would even work if the variation registration related calls came before the actual _block_ registration calls.

ockham replied:
Also, I don't think we have to check for the existence of `registerBlockVariation`, since on [R55](https://github.com/Automattic/jetpack/pull/17192/files#diff-02a96ee03b481186c273d2f9f1e49c09R55) we're already checking for `settings.variations`: If the `core/embed` block's `settings` has a `variations` field, then we can assume that `registerBlockVariation` also exists.

ockham commented on extensions/extended-blocks/core-embed/facebook.js:
```diff
@@ -0,0 +1,83 @@
+import { Path, SVG } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+const embedFacebookIcon = {
+	foreground: '#3b5998',
+	src: (
+		<SVG viewBox="0 0 24 24">
+			<Path d="M20 3H4c-.6 0-1 .4-1 1v16c0 .5.4 1 1 1h8.6v-7h-2.3v-2.7h2.3v-2c0-2.3 1.4-3.6 3.5-3.6 1 0 1.8.1 2.1.1v2.4h-1.4c-1.1 0-1.3.5-1.3 1.3v1.7h2.7l-.4 2.8h-2.3v7H20c.5 0 1-.4 1-1V4c0-.6-.4-1-1-1z" />
+		</SVG>
+	),
+};
```
Copied over from [Gutenberg](https://github.com/WordPress/gutenberg/blob/ed848a0ac49257c1c8162fef09440a36c48a8375/packages/block-library/src/embed/icons.js#L44-L51), in case it's removed there.

ockham commented on extensions/extended-blocks/core-embed/instagram.js:
```diff
@@ -0,0 +1,76 @@
+			<Path d="M12 4.622c2.403 0 2.688.01 3.637.052.877.04 1.354.187 1.67.31.42.163.72.358 1.036.673.315.315.51.615.673 1.035.123.317.27.794.31 1.67.043.95.052 1.235.052 3.638s-.01 2.688-.052 3.637c-.04.877-.187 1.354-.31 1.67-.163.42-.358.72-.673 1.036-.315.315-.615.51-1.035.673-.317.123-.794.27-1.67.31-.95.043-1.234.052-3.638.052s-2.688-.01-3.637-.052c-.877-.04-1.354-.187-1.67-.31-.42-.163-.72-.35 ...
+		</G>
+	</SVG>
+);
```
Copied over from [Gutenberg](https://github.com/WordPress/gutenberg/blob/ed848a0ac49257c1c8162fef09440a36c48a8375/packages/block-library/src/embed/icons.js#L52-L58), in case it's removed there.

ockham commented:
> I may be missing something here, but since the embed registration will also be removed from Core in [50861-core](https://href.li/?https://core.trac.wordpress.org/ticket/50861), won't we be creating issues on sites where Jetpack's Shortcodes module (where both Facebook and Instagram embeds are registered) is deactivated? On those sites, we'll show the 2 embed variations, but won't have the embed registration happening behind the scenes.
> 
> This may happen on a lot of sites since the module is not activated by default.

Fixed now. Added the following to testing instructions:

> - Remove all patches that you've applied on top of this branch, or on your sandbox. Disable your Jetpack test install's `shortcodes` module (`yarn docker:wp jetpack module deactivate shortcodes` if you're running the local Docker container). Reload the editor, and verify that the FB and IG embeds are gone from the inserter.

Should be ready for another look :blush:

jeherve commented on extensions/extended-blocks/core-embed/facebook.js:
```diff
@@ -0,0 +1,83 @@
+	name: 'facebook',
+	title: 'Facebook',
+	icon: embedFacebookIcon,
+	keywords: [ __( 'social' ) ],
+	description: __( 'Embed a Facebook post.' ),
```
```suggestion
	keywords: [ _x( 'social', 'block search term', 'jetpack' ) ],
	description: __( 'Embed a Facebook post.', 'jetpack' ),
```

jeherve commented on extensions/extended-blocks/core-embed/instagram.js:
```diff
@@ -0,0 +1,76 @@
+	name: 'instagram',
+	title: 'Instagram',
+	icon: embedInstagramIcon,
+	keywords: [ __( 'image' ), __( 'social' ) ],
+	description: __( 'Embed an Instagram post.' ),
```
```suggestion
	keywords: [ _x( 'image', 'block search term', 'jetpack' ), _x( 'social', 'block search term', 'jetpack' ) ],
	description: __( 'Embed an Instagram post.', 'jetpack' ),
```

jeherve commented on extensions/extended-blocks/core-embed/instagram.js:
```diff
@@ -0,0 +1,76 @@
+import { addFilter } from '@wordpress/hooks';
+import { registerBlockVariation, unregisterBlockVariation } from '@wordpress/blocks';
+import { G, Path, SVG } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
```
```suggestion
import { __, _x } from '@wordpress/i18n';
```

jeherve commented on extensions/extended-blocks/core-embed/facebook.js:
```diff
@@ -0,0 +1,83 @@
+import { addFilter } from '@wordpress/hooks';
+import { registerBlockVariation, unregisterBlockVariation } from '@wordpress/blocks';
+import { Path, SVG } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
```
```suggestion
import { __, _x } from '@wordpress/i18n';
```

jeherve commented on extensions/extended-blocks/core-embed/instagram.js:
```diff
@@ -0,0 +1,76 @@
+import { __ } from '@wordpress/i18n';
+
+const embedInstagramIcon = (
+	<SVG viewBox="0 0 24 24">
```
I think it would be nice to move those to `extensions/shared/icons.js`, with the rest of our icons.

ockham replied:
Right, Facebook is actually already there. I thought that maybe they looked a bit different, but the only visible difference (at that scale) seems to be blue vs black:

Before:
![image](https://user-images.githubusercontent.com/96308/94557582-0e4ce080-025f-11eb-93bc-3015b65ad8a1.png)


After:
![image](https://user-images.githubusercontent.com/96308/94557276-a39ba500-025e-11eb-893f-4809261ba275.png)

I guess the de-duplication outweighs the color change. Maybe we can even modify the component to allow it to inject additional props/colors later; or leave as-is to emphasise that this is different from the Core-provided embed.

I'll change :+1:

ockham replied:
Ah, interesting. When moving the Instagram icon, I noticed that we're already using one for the Lastest Insta Posts block:

![image](https://user-images.githubusercontent.com/96308/94559462-92a06300-0261-11eb-91c1-b5f18ba9138f.png)

I tracked it down to this code: https://github.com/Automattic/jetpack/blob/1527e0be4d25312d5fb4672a3f856e11ad6d84db/extensions/blocks/instagram-gallery/index.js#L21-L24

Turns out these are coming from [dashicons](https://developer.wordpress.org/resource/dashicons/#media-code), which have both an FB and an IG icon.

I'll try using those, and setting their color to JP green (to emphasize that they're coming from JP).

ockham replied:
d3df472:

![image](https://user-images.githubusercontent.com/96308/94559657-dabf8580-0261-11eb-989f-23fc7c0967c7.png)

![image](https://user-images.githubusercontent.com/96308/94559680-e448ed80-0261-11eb-9ff0-225c2bd286d9.png)

jeherve commented on _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php:
```diff
@@ -369,6 +369,8 @@ public function get_module( $request ) {
 				&& ( new Status() )->is_offline_mode()
 			) {
 				$module['activated'] = false;
+			} elseif ( Jetpack::is_module_active( $request['slug'] ) ) { // TODO: Move into `Jetpack::get_module()`?
```
Do we need to check that `slug` is set before we try to use it?

ockham replied:
I don't think we do, since we're already relying on it to be present a few lines above:

https://github.com/Automattic/jetpack/blob/1527e0be4d25312d5fb4672a3f856e11ad6d84db/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php#L360-L364

jeherve commented on _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php:
```diff
@@ -369,6 +369,8 @@ public function get_module( $request ) {
 				&& ( new Status() )->is_offline_mode()
 			) {
 				$module['activated'] = false;
+			} elseif ( Jetpack::is_module_active( $request['slug'] ) ) { // TODO: Move into `Jetpack::get_module()`?
```
This is an odd one, based on https://github.com/Automattic/jetpack/blob/e5a81d6e74516d2879f16caf9b0e2fac32bb16af/class.jetpack-admin.php#L136 I would have expected to have that information already?

ockham replied:
Yeah, I was surprised, too. (Forgot to leave a comment here, apologies.)

The line you found is used for the modules _list_ endpoint (`/modules/all`), but not for the individual one (`/modules/shortcodes`). (And I didn't want to use the list 'cause it's pretty big.)

**Here's the a call stack (of sorts) for the list endpoint:**

https://github.com/Automattic/jetpack/blob/1527e0be4d25312d5fb4672a3f856e11ad6d84db/_inc/lib/class.core-rest-api-endpoints.php#L276-L281
https://github.com/Automattic/jetpack/blob/1527e0be4d25312d5fb4672a3f856e11ad6d84db/_inc/lib/class.core-rest-api-endpoints.php#L74
https://github.com/Automattic/jetpack/blob/acfd44811fd55e3ead09648ab95adae92ce860b5/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php#L166
https://github.com/Automattic/jetpack/blob/acfd44811fd55e3ead09648ab95adae92ce860b5/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php#L177-L180
https://github.com/Automattic/jetpack/blob/acfd44811fd55e3ead09648ab95adae92ce860b5/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php#L192-L195
https://github.com/Automattic/jetpack/blob/acfd44811fd55e3ead09648ab95adae92ce860b5/class.jetpack-admin.php#L136

---

**Compare that to the individual module endpoint:**

https://github.com/Automattic/jetpack/blob/1527e0be4d25312d5fb4672a3f856e11ad6d84db/_inc/lib/class.core-rest-api-endpoints.php#L308-L312
https://github.com/Automattic/jetpack/blob/1527e0be4d25312d5fb4672a3f856e11ad6d84db/_inc/lib/class.core-rest-api-endpoints.php#L73
https://github.com/Automattic/jetpack/blob/1527e0be4d25312d5fb4672a3f856e11ad6d84db/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php#L322
https://github.com/Automattic/jetpack/blob/1527e0be4d25312d5fb4672a3f856e11ad6d84db/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php#L334-L338
https://github.com/Automattic/jetpack/blob/1527e0be4d25312d5fb4672a3f856e11ad6d84db/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php#L359-L362

`Jetpack::get_module()` doesn't add the `activated` field.

ockham replied:
As my `// TODO: Move into Jetpack::get_module()?` inline comment states, I was considering to move the `activated` field there. I'm not totally sure about the rationale behind decisions here; oversight aside, here are some hypotheses/considerations:

- `Jetpack::get_module()` should only expose 'static' information -- stuff that's encoded in the module's code, and doesn't change through user interaction.
- `Jetpack::get_module()` sits at a pretty low level. There might be unit or snapshot tests that could be easily thrown off by a newly added field.

Since I don't want this change to blow up the size of this PR (potentially needing to update tests, and research the full impact), I'm leaning towards keeping the change as-is, with the TODO as a reminder to potentially revisit.

retrofox commented on extensions/editor.js:
```diff
@@ -6,8 +6,8 @@ import './shared/block-category';
 import './shared/external-media';
-import './extended-blocks/paid-blocks';
 import './extended-blocks/core-embed';
+import './extended-blocks/paid-blocks';
```
minor suggestion: this change is not neccessary.

ockham replied:
Thanks for pointing it out! This happened when I was rebasing the PR. I'm thinking of keeping it, since it alphabetizes imports from the `./extended-blocks` folder neatly :sweat_smile:

ockham commented on extensions/extended-blocks/core-embed/facebook.js:
```diff
@@ -0,0 +1,81 @@
+import { getIconColor } from '../../shared/block-icons';
+
+const facebookVariation = {
+	// Deprecate Facebook Embed per FB policy
+	// See: https://developers.facebook.com/docs/plugins/oembed-legacy
```
Oh, I need to remove this.

```suggestion
```

retrofox commented on extensions/extended-blocks/core-embed/facebook.js:
```diff
@@ -0,0 +1,81 @@
+	},
+	keywords: [ _x( 'social', 'block search term', 'jetpack' ) ],
+	description: __( 'Embed a Facebook post.', 'jetpack' ),
+	patterns: [ /^https?:\/\/www\.facebook.com\/.+/i ],
```
I think `www` should be optional here, too.

ockham replied:
I copy-pasted the pattern from the [original GB embeds](https://github.com/WordPress/gutenberg/pull/24472/files#diff-d6e213880d8d5e45b0bf87afa293c01aL64) :thinking: Even though making the `www` optional makes sense to me, I'd rather not change the patterns TBH since I assume GB folks had them tried and tested :sweat_smile:

retrofox replied:
Ok, what I'm trying to raise up if you paste a link without the `www`, the variation won't be applied and it seems to be an issue to me.

ockham replied:
Agree -- it _is_ surprising. I'm really just trying to be as consistent as possible with Core here :slightly_smiling_face: Note that they _did_ have the `www` marked optional for Instagram (so it seems like a conscious decision _not_ to have it for Facebook): https://github.com/WordPress/gutenberg/pull/24472/files#diff-d6e213880d8d5e45b0bf87afa293c01aL77

Maybe FB now has canonical redirects from `www.facebook.com` to `facbook.com`? Anyway, let's investigate later :+1:

ockham commented:
Addressed feedback; should be ready for another look :smile:

ockham commented:
Per p1601385813135300-slack-CBG1CP4EN, we've decided to decouple the FB and IG embeds from the `shortcodes` module, and enabled them on the sole condition that JP is connected to WP.com. (The reason for this is that embeds will soon need their API requests authenticated with a token, and that token is stored on WP.com.)

Setting this PR to Blocked while I'm filing a separate PR for the decoupling.

ockham commented:
> Per [p1601385813135300-slack-CBG1CP4EN](https://href.li/?p1601385813135300-slack-CBG1CP4EN), we've decided to decouple the FB and IG embeds from the `shortcodes` module, and enabled them on the sole condition that JP is connected to WP.com. (The reason for this is that embeds will soon need their API requests authenticated with a token, and that token is stored on WP.com.)
> 
> Setting this PR to Blocked while I'm filing a separate PR for the decoupling.

:point_right: https://github.com/Automattic/jetpack/pull/17301
Now going to modify this PR accordingly.

ockham commented:
Highlighting here that apparently we _don't_ need an extra check to make sure we're connected for the IG embed variation, since our blocks (and variations) are only available if JP is connected anyway.

ockham commented:
Should be ready for another look. #17301 should be merged before this (to make sure the embeds don't break if the `shortcodes` module is disabled), but that's not a blocker for testing/reviewing this PR.

jeherve reviewed (changes requested):
About f7fcb813b798855516776c3ae88cd3c3f73c6c6d: we do in fact need that check, as blocks and those embed variations can be available on sites that are not connected to WordPress.com yet, but that use [Offline mode](https://jetpack.com/support/development-mode/). In those cases, we don't want the Instagram variation to be available.

ockham commented:
> About [f7fcb81](https://github.com/Automattic/jetpack/commit/f7fcb813b798855516776c3ae88cd3c3f73c6c6d): we do in fact need that check, as blocks and those embed variations can be available on sites that are not connected to WordPress.com yet, but that use [Offline mode](https://jetpack.com/support/development-mode/). In those cases, we don't want the Instagram variation to be available.

Okay, I'll revert!

ockham commented:
Rebased.

ockham commented:
> > About [f7fcb81](https://github.com/Automattic/jetpack/commit/f7fcb813b798855516776c3ae88cd3c3f73c6c6d): we do in fact need that check, as blocks and those embed variations can be available on sites that are not connected to WordPress.com yet, but that use [Offline mode](https://jetpack.com/support/development-mode/). In those cases, we don't want the Instagram variation to be available.
> 
> Okay, I'll revert!

Done in b69ecbc.

jeherve reviewed (approved):
This tests well for me. Merging.

jeherve commented:
Cherry-picked to `branch-9.0` in f1cd0703638494ae96045c9bebdde6af8f6f2ede

I'll let you deal with the WordPress.com counterpart?

ockham commented:
> Cherry-picked to `branch-9.0` in [f1cd070](https://github.com/Automattic/jetpack/commit/f1cd0703638494ae96045c9bebdde6af8f6f2ede)

Thanks @jeherve !

> I'll let you deal with the WordPress.com counterpart?

Will do! It's kinda blocked by D49389-code, but I'll work around that.
